### PR TITLE
Add support for clean Avro issue#643

### PR DESCRIPTION
**Pull request by tagarr**
Add option to use headers instead of magic byte for serializing the global-ids

### CODE REVIEW

File: utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaDeserializer.java [hunk 1]
```diff
@@ -76,4 +82,40 @@ public U deserialize(String topic, byte[] data) {
...
+            throw new RuntimeException("ArtifactId not found in headers.");
+        }
+        if (version == null) {
+            ArtifactMetaData amd = getClient().getArtifactMetaData(artifactId);
+            return amd.getGlobalId();
+        } else {
+            VersionMetaData vmd = getClient().getArtifactVersionMetaData(version, artifactId);
+            return vmd.getGlobalId();
+        }
+    }
```
Comment (EricWittmann) [inline utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaDeserializer.java:120]: Some of this logic is duplicated in `JsonSchemaKafkaDeserializer` I think - is there an opportunity to consolidate this?

File: utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaSerDe.java [hunk 2]
```diff
@@ -48,12 +51,19 @@
     public static final String REGISTRY_ID_HANDLER_CONFIG_PARAM = "apicurio.registry.id-handler";
     public static final String REGISTRY_CONFLUENT_ID_HANDLER_CONFIG_PARAM = "apicurio.registry.as-confluent";
 
+    // Constants for using headers to store the ids
+    public static final String USE_HEADERS = "apicurio.use.headers";
```
Comment (EricWittmann) [inline utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaSerDe.java:55]: I think this should be `apicurio.registry.use.headers` to match the other config prop names.

File: utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaSerializer.java [hunk 3]
```diff
@@ -65,12 +74,18 @@ public AbstractKafkaSerializer(
...
-            getIdHandler().writeId(id, out);
-            serializeData(schema, data, out);
+            if (headerUtils != null) {
+                headerUtils.addSchemaHeaders(headers, artifactId, id);
+                serializeData(headers, schema, data, out);
+            } else {
+                out.write(MAGIC_BYTE);
+                getIdHandler().writeId(id, out);
+                serializeData(schema, data, out);
+            }
```
Comment (EricWittmann) [inline utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaSerializer.java:84]: I was expecting the decision point to be a configuration property, not the presence of the `Headers` object - what determines which variant of `serialize` is called?

File: utils/serde/src/main/java/io/apicurio/registry/utils/serde/AvroKafkaDeserializer.java [hunk 4]
```diff
@@ -77,8 +82,24 @@ protected Schema toSchema(Response response) {
...
+        }
+        if (encoding == null) {
+            // no encoding in header or no headers so use config
+            encoding = configEncoding;
+        }
         try {
             DatumReader<U> reader = avroDatumProvider.createDatumReader(schema);
+            if (encoding == null) {
+                encoding = configEncoding;
+            }
```
Comment (EricWittmann) [inline utils/serde/src/main/java/io/apicurio/registry/utils/serde/AvroKafkaDeserializer.java:102]: Isn't this caught above on lines 94-97?

File: utils/serde/src/main/java/io/apicurio/registry/utils/serde/AvroKafkaSerializer.java [hunk 5]
```diff
@@ -104,6 +112,12 @@ protected void serializeData(Schema schema, U data, OutputStream out) throws IOE
         encoder.flush();
     }
 
+    @Override
+    protected void serializeData(Headers headers, Schema schema, U data, ByteArrayOutputStream out) throws IOException {
+        headerUtils.addEncodingHeader(headers, encoding);
```
Comment (EricWittmann) [inline utils/serde/src/main/java/io/apicurio/registry/utils/serde/AvroKafkaSerializer.java:117]: Might this be `null` sometimes?

**Comment by EricWittmann** [comment]
Overall this makes sense - the thing I'm confused about (in general) is that I thought whether or not the Kafka `Headers` option was present simply depended on which variant of `deserialize()` you implemented on `org.apache.kafka.common.serialization.Deserializer`.  This PR seems to sometimes create a new instance of `Headers` and also seems to make decisions based on whether the headers is `null` or not.  That said, I might have missed the overall approach due to the fragmented nature of doing code reviews the GitHub way. :)

File: utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaSerDe.java (hunk 2 above)
Comment (tagarr) [reply utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaSerDe.java:55]: Agree

File: utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaSerializer.java (hunk 3 above)
Comment (tagarr) [reply utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaSerializer.java:84]: The headerUtil is effectively only created if the config has been set

File: utils/serde/src/main/java/io/apicurio/registry/utils/serde/AvroKafkaDeserializer.java (hunk 4 above)
Comment (tagarr) [reply utils/serde/src/main/java/io/apicurio/registry/utils/serde/AvroKafkaDeserializer.java:102]: yes

File: utils/serde/src/main/java/io/apicurio/registry/utils/serde/AvroKafkaSerializer.java (hunk 5 above)
Comment (tagarr) [reply utils/serde/src/main/java/io/apicurio/registry/utils/serde/AvroKafkaSerializer.java:117]: yes

**Comment by tagarr** [comment]
So the main decision point is in the producer, which will either put the global id in the header or after a magic byte. The consumer can then just check for presence of magic byte. If it is there, then use normal deserialization otherwise look in the header, this means that a consumer will be able to consume messages from a topic from producers that use either method

**Comment by EricWittmann** [comment]
Thanks for the high level explanation.  It makes sense.

**Review by EricWittmann — APPROVED** [review]

**Comment by EricWittmann** [comment]
Documented here: https://github.com/Apicurio/apicurio-registry/pull/957

**Pull request merged**